apps/emqx/test/emqx_common_test_helpers.erl

```diff
@@ -503,3 +518,247 @@ ensure_dashboard_listeners_started(emqx_dashboard) ->
+            {monitor_master, true},
+            {init_timeout, 10000},
+            {startup_timeout, 10000},
+            {erl_flags, erl_flags()}
+        ]
+    ),
+
+    pong = net_adm:ping(Node),
+    setup_node(Node, Opts),
+    Node.
+
+%% Node stopping
+stop_slave(Node) ->
+    ct_slave:stop(Node).
+
+%% EPMD starting
+start_epmd() ->
+    [] = os:cmd("\"" ++ epmd_path() ++ "\" -daemon"),
+    ok.
+
+epmd_path() ->
+    case os:find_executable("epmd") of
+        false ->
+            ct:pal(critical, "Could not find epmd.~n"),
+            exit(epmd_not_found);
+        GlobalEpmd ->
+            GlobalEpmd
+    end.
+
+%% Node initialization
+
+setup_node(Node, Opts) when is_list(Opts) ->
+    setup_node(Node, maps:from_list(Opts));
+setup_node(Node, Opts) when is_map(Opts) ->
+    %% Default base port is selected upon Node from 1100 to 65530 with step 10
+    BasePort = maps:get(base_port, Opts, 1100 + erlang:phash2(Node, 6553 - 110) * 10),
+    Apps = maps:get(apps, Opts, []),
+    StartApps = maps:get(start_apps, Opts, true),
+    JoinTo = maps:get(join_to, Opts, undefined),
+    EnvHandler = maps:get(env_handler, Opts, fun(_) -> ok end),
+    ConfigureGenRpc = maps:get(configure_gen_rpc, Opts, true),
+    LoadSchema = maps:get(load_schema, Opts, true),
+    LoadApps = maps:get(load_apps, Opts, [gen_rpc, emqx, ekka, mria] ++ Apps),
+    Env = maps:get(env, Opts, []),
+    Conf = maps:get(conf, Opts, []),
+    ListenerPorts = maps:get(listener_ports, Opts, [
+        {Type, listener_port(BasePort, Type)}
+     || Type <- [tcp, ssl, quic, ws, wss]
+    ]),
+
+    %% Load env before doing anything to avoid overriding
+    [ok = rpc:call(Node, application, load, [App]) || App <- LoadApps],
+
+    %% Needs to be set explicitly because ekka:start() (which calls `gen`) is called without Handler
+    %% in emqx_common_test_helpers:start_apps(...)
+    ?ON_TRUE(
+        ConfigureGenRpc,
+        begin
+            ok = rpc:call(Node, application, set_env, [
+                gen_rpc, tcp_server_port, gen_rpc_port(BasePort)
+            ]),
+            ok = rpc:call(Node, application, set_env, [gen_rpc, port_discovery, manual])
+        end
+    ),
+
+    %% Here we start the apps
+    EnvHandlerForRpc =
+        fun(App) ->
+            %% We load configuration, and than set the special enviroment variable
+            %% which says that emqx shouldn't load configuration at startup
+            %% Otherwise, configuration get's loaded and all preset env in envhandler is lost
+            ?ON_TRUE(
+                LoadSchema,
+                begin
+                    emqx_config:init_load(emqx_schema),
+                    application:set_env(emqx, init_config_load_done, true)
+                end
+            ),
+
+            %% Need to set this otherwise listeners will conflict between each other
+            [
+                ok = emqx_config:put([listeners, Type, default, bind], {
+                    {127, 0, 0, 1}, Port
```

Comment:
I think a much better approach would be to bind unique IP addresses to each node (e.g. node 3 -> `{127, 0, 0, 3}` ) to the nodes, and keep the default ports.

Reply:
Ports are unique, and collision can happen when using unique IPs too. For example, when multiple multi-node tests are started simultaneously. There's no difference between these two solutions.

Reply:
I think using IPs is a more idiomatic way, like one IP to one node relation. But there's no practical difference to be sure

Reply:
It took me some time to reimplement it and I've found out that by default master node starts all listeners on 0.0.0.0, and this will require to rewrite a lot of tests to use 127.0.0.1 as a default interface.

Perhaps you know some easy magic way to set this address as a default for all tests.

Reply:
> I think using IPs is a more idiomatic way, like one IP to one node relation. But there's no practical difference to be sure

It makes the test setup easier to understand. Therefore, debugging failing testcases is easier. This is important for overall development speed in the project.

> It took me some time to reimplement it and I've found out that by default master node starts all listeners on 0.0.0.0, and this will require to rewrite a lot of tests to use 127.0.0.1 as a default interface.

Is it possible to change the default config?

Reply:
Yeah, it is possible to do it during tests, but this is very tricky. It'll look like this.
```erlang
emqx_config:init_load(emqx_schema),
application:set_env(emqx, init_config_load_done, true)
[ok = emqx_config:put([listeners, Type, default, bind], {Host, Port}) || Type <- Types]
```

To change this by default for all tests, we'd need some separate `emqx.conf` for testing